Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
-		hitMOID = g_SceneMan.GetMOIDPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
-		// Don't count MOIDs ignored
-		if (hitMOID != g_NoMOID && !atom->IsIgnoringMOID(hitMOID))
-		{
-			// Save the correct MOID to search for other atom intersections with
-			currentMOID = hitMOID;
-
-			// Get the MO we seem to be intersecting
-			MovableObject *tempMO = g_MovableMan.GetMOFromID(hitMOID);
-			RTEAssert(tempMO, "Intersected MOID couldn't be translated to a real MO!");
-			tempMO = tempMO->GetRootParent();
-
-			if (tempMO->GetsHitByMOs())
-			{
-				// Make that MO draw itself again in the MOID layer so we can find its true edges
-				pIntersectedMO = tempMO;
-				pIntersectedMO->Draw(g_SceneMan.GetMOIDBitmap(), Vector(), g_DrawMOID, true);
-				break;
-				// TODO: Remove
-				//                g_FrameMan.SaveBitmapToBMP(g_SceneMan.GetMOIDBitmap(), "MOIDMap");
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool AtomGroup::ResolveTerrainIntersection(Vector &position, unsigned char strongerThan) const {
+		std::list<Atom *> intersectingAtoms;
+		MOID hitMaterial = g_MaterialAir;
+
+		float strengthThreshold = (strongerThan != g_MaterialAir) ? g_SceneMan.GetMaterialFromID(strongerThan)->GetIntegrity() : 0.0F;
+		bool rayHit = false;
+
+		Vector atomOffset = Vector();
+		Vector atomPos = Vector();
+
+		// First go through all Atoms to find the first intersection and get the intersected MO
+		for (Atom *atom : m_Atoms) {
+			atomOffset = m_OwnerMO->RotateOffset(atom->GetOffset());
+			atom->SetupPos(position + atomOffset);
+			atomPos = atom->GetCurrentPos();
+			hitMaterial = g_SceneMan.GetTerrain()->GetPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
+			if (hitMaterial != g_MaterialAir && strengthThreshold > 0.0F && g_SceneMan.GetMaterialFromID(hitMaterial)->GetIntegrity() > strengthThreshold) {
+				intersectingAtoms.push_back(atom);
 			}
 		}
-    }
+		if (intersectingAtoms.empty()) {
+			return true;
+		}
 
-    // If no intersected MO is found, we are clear and can quit
-	if (!pIntersectedMO) {
-		return false;
-	}
+		// If all Atoms are intersecting, we're screwed?!
+		if (intersectingAtoms.size() >= m_Atoms.size()) {
+			return false;
+		}
+
+		Vector exitDirection = Vector();
+
+		// Go through all intersecting Atoms and find their average inverse normal
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			exitDirection += m_OwnerMO->RotateOffset(intersectingAtom->GetNormal());
+		}
+
+		// We don't have a direction to go, so quit
+		// TODO: Maybe use previous position to create an exit direction instead then?
+		if (exitDirection.IsZero()) {
+			return false;
+		}
+
+		// Invert and set appropriate length
+		exitDirection = -exitDirection;
+		exitDirection.SetMagnitude(m_OwnerMO->GetDiameter());
+
+		// See which of the intersecting Atoms has the longest to travel along the exit direction before it clears
```

Comment:
Seems like this is worth pulling out into a private method instead of needing this comment.